### PR TITLE
Spec and test w,h constraints for BPTC, RGTC, and S3TC.

### LINKED ISSUE #3117: Relax implicit power-of-two restrictions for mipmapped textures in WEBGL_compressed_texture_s3tc
Context:
https://github.com/BinomialLLC/basis_universal/issues/154
https://github.com/KhronosGroup/KTX-Software/issues/254
https://github.com/KhronosGroup/KTX-Specification/issues/123

https://bugs.chromium.org/p/angleproject/issues/detail?id=4841
https://www.khronos.org/registry/webgl/extensions/WEBGL_compressed_texture_s3tc/
https://www.khronos.org/registry/OpenGL/extensions/ANGLE/ANGLE_texture_compression_dxt.txt

It appears that longstanding restrictions in the ANGLE_texture_compression_dxt extension, which were lifted into WebGL's WEBGL_compressed_texture_s3tc extension, have had the following consequences:

1) Because all texture levels beyond those of width/height 1 or 2 have to have width/height divisible by 4, this implies that all levels have to be power-of-two sized.

2) Because this restriction exists for S3TC, and because this is one possible transcoding target of Basis Universal textures, this implies that all textures processed with this pipeline have to obey the power-of-two restriction in case they are transcoded at runtime to S3TC.

The WebGL WG must figure out a solution to this problem. https://bugs.chromium.org/p/angleproject/issues/detail?id=4841
 tracks investigation at the ANGLE layer, which will inform a specification-level solution.

### PR DESCRIPTION
**Pull request by jdashg**
`w << level` and `h << level` must be multiples of 4.
This way, we forbid image allocations that are never valid, such as
1x1 for level 1. (would only be valid with a 2x2 base level, which is
forbidden)

Resolves #3117.

### CODE REVIEW

File: extensions/EXT_texture_compression_bptc/extension.xml [hunk 1]
```diff
@@ -64,7 +64,7 @@
           <dt>COMPRESSED_RGB_BPTC_SIGNED_FLOAT_EXT</dt>
           <dt>COMPRESSED_RGB_BPTC_UNSIGNED_FLOAT_EXT</dt>
           <dd>
-          <p>When <code>level</code> equals zero <code>width</code> and <code>height</code>
+          <p><code>width &lt;&lt; level</code> and <code>height &lt;&lt; level</code>
           must be a multiple of 4.</p>
```
Comment (kenrussell) [inline extensions/EXT_texture_compression_bptc/extension.xml:68]: Here and throughout: I'm not sure this encodes the constraints discussed in #3117. Rereading the discussion there, it sounds like two branches are needed.
 - For both WebGL 1.0 and 2.0 contexts, the base level's width and height must each be a multiple of 4.
 - For WebGL 1.0 contexts, the base level's width and height must each be a power of two.

This should apply to all of these texture formats. DXT in particular needs to retain the power-of-two restriction for WebGL 1.0 contexts. WebGL's BPTC and RGTC extensions are written against the 1.0 spec even though WebGL 1.0 doesn't have RG textures, so theoretically they should retain the same restrictions.

File: sdk/tests/conformance/extensions/ext-texture-compression-bptc.html [hunk 2]
```diff
@@ -87,10 +87,16 @@
...
+        expectation: gl.INVALID_OPERATION, message: "implied base mip 2x4 is invalid" },
+      { level: 1, width: 2, height: 1,
+        expectation: gl.INVALID_OPERATION, message: "implied base mip 4x2 is invalid" },
       { level: 1, width: 2, height: 2,
-        expectation: gl.NO_ERROR, message: "level > 0, is valid" },
-      { level: 2, width: 1, height: 1,
-        expectation: gl.NO_ERROR, message: "level > 0, is valid" },
+        expectation: gl.NO_ERROR, message: "implied base mip 4x4 is valid" },
+      { level: 2, width: 1, height: 3,
+        expectation: gl.NO_ERROR, message: "implied base mip 4x12 is valid" },
```
Comment (kenrussell) [inline sdk/tests/conformance/extensions/ext-texture-compression-bptc.html:99]: I think the base mip of 4x12 should only be valid in a WebGL 2.0 context.

File: sdk/tests/conformance/extensions/ext-texture-compression-rgtc.html [hunk 3]
```diff
@@ -92,10 +92,16 @@
...
+        expectation: gl.INVALID_OPERATION, message: "implied base mip 2x4 is invalid" },
+      { level: 1, width: 2, height: 1,
+        expectation: gl.INVALID_OPERATION, message: "implied base mip 4x2 is invalid" },
       { level: 1, width: 2, height: 2,
-        expectation: gl.NO_ERROR, message: "level > 0, is valid" },
-      { level: 2, width: 1, height: 1,
-        expectation: gl.NO_ERROR, message: "level > 0, is valid" },
+        expectation: gl.NO_ERROR, message: "implied base mip 4x4 is valid" },
+      { level: 2, width: 1, height: 3,
+        expectation: gl.NO_ERROR, message: "implied base mip 4x12 is valid" },
```
Comment (kenrussell) [inline sdk/tests/conformance/extensions/ext-texture-compression-rgtc.html:104]: Similarly here, I think this is only valid in a WebGL 2.0 context.

**Review by kenrussell — COMMENTED** [review]
Thanks for revisiting and improving this. A few comments.

**Comment by kenrussell** [comment]
@lexaknyazev could you also please take a look at these changes?

File: extensions/EXT_texture_compression_bptc/extension.xml (hunk 1 above)
Comment (jdashg) [reply extensions/EXT_texture_compression_bptc/extension.xml:68]: I'm not really convinced by the webgl1 pot requirement, which I don't think in enforced or tested, and isn't spec'd. Patches welcome but I don't want to tackle that here.

Comment (kenrussell) [reply extensions/EXT_texture_compression_bptc/extension.xml:68]: OK, I see, even though we discussed this in the earlier issue it wasn't enforced. These changes are fine then.

File: sdk/tests/conformance/extensions/ext-texture-compression-rgtc.html (hunk 3 above)
Comment (kenrussell) [reply sdk/tests/conformance/extensions/ext-texture-compression-rgtc.html:104]: Never mind, per above.

File: sdk/tests/conformance/extensions/ext-texture-compression-bptc.html (hunk 2 above)
Comment (kenrussell) [reply sdk/tests/conformance/extensions/ext-texture-compression-bptc.html:99]: Never mind, this wasn't enforced previously.

**Comment by kenrussell** [comment]
Let's go ahead with these changes. @lexaknyazev if you have concerns we can address them in follow-on changes.

**Pull request merged**